### ISSUE TITLE
Cypress’ parallel mode overwrites files among Cypress instances 

### ISSUE THREAD

**Issue by romankhomitskyi** [cypress-io/cypress#26211]
### Current behavior

My Cypress test specs are coded in TypeScript. Compiling TypeScript code in Cypress’ parallel mode causes overwriting in files among Cypress instances 

I am getting the following errors on CI
- Invalid or unexpected token
- Unexpected end of input
- Failed to execute 'atob' on 'Window'
- is not a function, however it is a function and works OK on local and without parallel mode, or with work around

![Screenshot 2023-03-24 at 10 19 13](https://user-images.githubusercontent.com/57874507/227463579-c73df059-2051-4113-a153-e4355611d18f.png)

![Screenshot 2023-03-24 at 10 20 10](https://user-images.githubusercontent.com/57874507/227463764-c2c8267f-134f-4584-a682-d0281e8a4189.png)

![Screenshot 2023-03-24 at 10 20 42](https://user-images.githubusercontent.com/57874507/227463872-79ec6424-1019-4a10-89f8-77710b6a6997.png)

![Screenshot 2023-03-24 at 10 21 24](https://user-images.githubusercontent.com/57874507/227464004-d524edae-0ddd-42ce-8ac0-ca4c35973716.png)

![Screenshot 2023-03-24 at 10 21 43](https://user-images.githubusercontent.com/57874507/227464168-fc36233b-5efc-436b-af9d-9c88c6fa5b6e.png)

And also this error in this issue https://github.com/cypress-io/cypress/issues/9442

The only workaround is to isolate test files in their test runs, so test code will not get overwritten by other instances

I have this code to append unique IDs to filenames before Webpack processes test files in `cypress.config.ts`

```
import { v4 as uuid } from "uuid";

import webpackPreprocessor from "@cypress/webpack-batteries-included-preprocessor";

const filePreprocessor = webpackPreprocessor({ typescript: "typescript" });

const id = uuid();
on("file:preprocessor", (file) => {
  file.outputPath = file.outputPath.replace(
    /^(.*\/)(.*?)(\..*)$/,
    `$1$2.${id}$3`
  );
  console.log("File output", file.outputPath);
  return filePreprocessor(file);
});

```

The path will be replaced with `“specs/my_feature/abc.<uuid>.spec.ts”. `<uuid> is unique among each cypress instances.

### Desired behavior

Cypress should not overwrite  files among Cypress instances
It should be implemented under the hood 

### Test code to reproduce

I run 58 cypress instances in one container which is starting in one Gitlab job

Overall I have the following flow

4 parallel gitlab jobs
4 cypress docker container in each
58 cypress instances in each container

I am using AWS spot instances

![Screenshot 2023-03-24 at 10 31 13](https://user-images.githubusercontent.com/57874507/227466164-4dea8078-9fbd-4709-ad7c-5e867597072f.png)

### Cypress Version

12.5.0

### Node version

18.13.0

### Operating System

Debian, cypress/factory docker image

**Comment by AtofStryker** [comment]
Hi @romankhomitskyi . Thank you for opening an issue. This seems like it might be a hard issue to reproduce. Is there anyway we can get a small scale reproduction repository going that shows the issues with the file overwriting with the preprocessor?  Also what version of Typescript are you using?

**Comment by romankhomitskyi** [comment]
> Hi @romankhomitskyi . Thank you for opening an issue. This seems like it might be a hard issue to reproduce. Is there anyway we can get a small scale reproduction repository going that shows the issues with the file overwriting with the preprocessor? Also what version of Typescript are you using?

Typescript: 4.9.5, and oldest


Here is an article describing a problem and solution to this issue

https://link.medium.com/Q9YrwZGaryb


Also @BlueWinds confirmed that is a root cause 
![SmartSelect_20230324_220014_Chrome](https://user-images.githubusercontent.com/57874507/227627916-edd0f11c-f6ce-4123-887a-d2934e6c90e6.jpg)


https://github.com/cypress-io/cypress/issues/9442

**Comment by romankhomitskyi** [comment]
The reproduction is a running multiple cypress runner instances on the same machine, which is not recommended  but anyway, they should not mutate each other

**Comment by emilyrohrbough** [comment]
@romankhomitskyi Yeah, Cypress is not setup out-of-the box to run in parallel on the same machine. The webpack output is assuming a single directory when leveraging the `cypress/webpack-preprocessor` and the way we load these processed files in, we assume the are already unique. The solution you posted is the correct path forward. I will flag as an `enhacement` for our team to consider.
